subcommittee/coding-guidelines/initiatives/safe-use-of-unsafe-guidelines/rust-embedded-lld-safe-definition.md

```diff
@@ -0,0 +1,295 @@
+propose a solution.
+
+Before to proceed to this analysis we need to understand what does mean
+`safe` for Rust language:
+
+## What does mean safe for Rust language.
+
+Rust is considered a safe language and the terms `safe` and `unsafe` are
+widely used in its documentation.
+
+What does mean `safe` in the context of Rust? I was not able to find a
+good definition of what is safe in Rust but there is quite clear
+definition of what shall be considered `unsafe` in Rust.
+
+In Rust Ferrocene specification is written the following [^1]:
+
+*Unsafety is the presence of unsafe operations in program text.*
+
+*An unsafe operation is an operation that may result in undefined
+behavior that is not diagnosed as a static error. Unsafe operations are
+referred to as unsafe Rust.*
+
+This definition is aligned with what is written in Rustnomicon[^2]:
+
+*No matter what, Safe Rust can't cause Undefined Behavior.*
+
+Moreover, the keyword `unsafe` is used in multiple context[^3]:
+
+  - *unsafe fn*: calling this function means abiding by a contract the
+    compiler cannot enforce.
+
+  - *unsafe trait*: implementing the trait means abiding by a contract
+    the compiler cannot enforce.
+
+  - *unsafe {}*: the contract necessary to call the operations inside
+    the block has been checked by the programmer and is guaranteed to be
+    respected.
+
+  - *unsafe impl*: the contract necessary to implement the trait has
+    been checked by the programmer and is guaranteed to be respected.
+
+In this discussion only unsafe blocks and unsafe functions. Unsafe
+functions can be called only in unsafe blocks and the user shall ensure
+to use unsafe functions fulfilling all requirements described in the
+documentation to avoid undefined behavior.
+
+This simplifies code review and bug identification. The developers shall
+inspect first unsafe blocks in case application shows an undefined
+behavior. However, this doesn`t exclude that a function developer forgot
+to mark a function as “unsafe”.
+
+### What is “undefined behavior” in Rust? 
+
+Ferrocene specification defines it as “unbounded error”[^4] or in other
+words: everything could happen.
+
+Rust Reference manual provide a non-exhaustive list of “undefined
+behaviors”[^5] The following behavior are worth to mentioned because it
+involves hardware:
+
+  - *Executing code compiled with platform features that the current
+    platform does not support (see target\_feature), except if the
+    platform explicitly documents this to be safe.*
+
+Before to discuss safe/unsafe in Low Level Drivers let`s see the meaning
+of “undefined behavior” in the context of microcontroller/microprocessor
+hardware.
+
+## Hardware undefined behavior
+
+In many data sheet or user manual of micro-controller and microprocessor
+is used the term *undefined behavior* and in these documents is
+described how to avoid the conditions that could generate an undefined
+behavior.
```

Comment:
I wasn't quite sure what this paragraph get getting at. Does this suggestion make sense, or are you trying to make a different point?

```suggestion
In the data sheets or user manuals of many micro-controllers and 
microprocessors the term *undefined behavior* is used mostly to describe
how to avoid the conditions that could generate an undefined behavior.
```